### DISCUSSION TITLE
Urgent: Swoole\Http\Client vs Swoole\Coroutine\HTTP\Client

### DISCUSSION THREAD

**Discussion by fakharksa** [swoole/swoole-src#4887] [Q&A] [answered]
Question-1: Why description of  "Swoole\Http\Client" is not available on official documentation for OpensWoole ? (official documentation describes only Coroutine version "Swoole\Coroutine\HTTP\Client").

Question-2: The get() function on Swoole\Http\Client takes two parameters the second of which is a callback / anonymous function, but this second parameter is not available on coroutine version of get(), why ? 

Question-3: The link https://openswoole.com/docs/modules/swoole-coroutine-http-client-get recieves body data using $client->body which dows not work and which is not available in official documentation, instead getBody() works and is available on Official Documentation. May be you need to update example on the link above.

**Comment by volodymyr-hordiienko** [comment]
As I understand openswoole is a fork , and looks like not supported a lot, so openswoole documentation may be not relevant for swoole project, correct me pls if I wrong

**Reply by fakharksa** [reply]
@Brightlycolor
No, that is not the case. The only detailed documentation available for Swoole is "openswoole.com". Other important facts to note are:

1. The author "Bruce Dow" of the book "Mastering Swoole PHP" is one of the key and main developer of Swoole, and he is the one who actually forked OpenSwoole from Swoole (As i read somewhere reason to fork was to add some security features quickly, as the PullRequest of the security-features in Swoole remained pending for some time, so Bruce Dow decided to make a fork and add security features).

2. The key / active developers of Swoole like "matyhtf" are also among the key contributors of OpenSwoole.

3. OpenSwoole is a superset of Swoole in terms of Features and Security. For example OpenSwoole implements gRPC, which is not available in Swoole.

4. OpenSwoole is the name which better describe the opensource nature of the project "Swoole".

5. Development on OpenSwoole Repo is more fast / active than on Swoole.

I strongly recommend the use of OpenSwoole over Swoole also because in my opinion "The one who holds the online-documentation / e-book is the one who wins more developers".

**Reply by volodymyr-hordiienko** [reply]
I have one another documentation here, https://wiki.swoole.com/#/ not sure its supported but very informative, i use swoole currently because its supported by hyperf framework. But truly i did not research which one better

**Reply by volodymyr-hordiienko** [reply]
Also, grpc is available for me with hyperf/swoole, not sure its supported natively or built by hyperf team based on tcp server, need to research

**Comment by NathanFreeman** [comment, marked as answer]
@fakharksa Hi.
Swoole\Http\Client is only available in the old version and it was renamed to Swoole\Coroutine\HTTP\Client in later versions.
Some functions of Swoole\Http\Client were changed in later versions.

I think you are confusing the github repo between the swoole and openswoole, here is swoole github repo.
Swoole and openswoole are different repo and the detailed documentation available for Swoole is https://www.swoole.com.

Swoole developers are still actively maintaining swoole, you can see the prs https://github.com/swoole/swoole-src/pulls , commits https://github.com/swoole/swoole-src/commits/master and releases https://github.com/swoole/swoole-src/releases

Swoole also has grpc too https://github.com/swoole/grpc

matyhtf is the key and active contributor to swoole rather than openswoole.

**Reply by fakharksa** [reply]
> @fakharksa Hi. Swoole\Http\Client is only available in the old version and it was renamed to Swoole\Coroutine\HTTP\Client in later versions. Some functions of Swoole\Http\Client were changed in later versions.
> 
> I think you are confusing the github repo between the swoole and openswoole, here is swoole github repo. Swoole and openswoole are different repo and the detailed documentation available for Swoole is https://www.swoole.com.
> 
> Swoole developers are still actively maintaining swoole, you can see the prs https://github.com/swoole/swoole-src/pulls , commits https://github.com/swoole/swoole-src/commits/master and releases https://github.com/swoole/swoole-src/releases
> 
> Swoole also has grpc too https://github.com/swoole/grpc
> 
> matyhtf is the key and active contributor to swoole rather than openswoole.

@brightlycolor (gRPC is available in Swoole natively, hyperf is just another MVC-Design-Pattern Framework wrapping MVC-related things around Asynchrony and multi-Processing provided by Swoole. Everything that "hyperf" exposes (other than MVC-related Design Patterns / Functionality) belongs to Swoole. For example, Model-View-Controller, Routers, Migrations do not belong to Swoole, but Asynchrony (Coroutine, Coroutine Clients, Asynchronous Servers, Task Workers (for long running tasks), Process and Inter-Process Communication, Process Pool Manager, Connection Pools, Channel, Network Protocols like TCP, UDP, HTTP 1, 1.1, 2, gRPC, RPC, Socket API, as well as Events, millisecond Timer, Request and Response objects, Server Configurations, Coroutine configurations etc. all come from Swoole).
